File: demos/launcher.c
/*
 * Copyright (C) 2008 Pierre-Luc Beaudoin <pierre-luc@pierlux.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <champlain/champlain.h>
#include "markers.h"

#define PADDING 10

static gboolean
map_view_button_release_cb (G_GNUC_UNUSED ClutterActor *actor,
    ClutterButtonEvent *event,
    ChamplainView *view)
{
  gdouble lat, lon;

  if (event->button != 1 || event->click_count > 1)
    return FALSE;

  lon = champlain_view_x_to_longitude (view, event->x);
  lat = champlain_view_y_to_latitude (view, event->y);

  g_print ("Map clicked at %f, %f \n", lat, lon);

  return TRUE;
}


static gboolean
zoom_in (G_GNUC_UNUSED ClutterActor *actor,
    G_GNUC_UNUSED ClutterButtonEvent *event,
    ChamplainView *view)
{
  champlain_view_zoom_in (view);
  return TRUE;
}


static gboolean
zoom_out (G_GNUC_UNUSED ClutterActor *actor,
    G_GNUC_UNUSED ClutterButtonEvent *event,
    ChamplainView *view)
{
  champlain_view_zoom_out (view);
  return TRUE;
}


static ClutterActor *
make_button (char *text)
{
  ClutterActor *button, *button_bg, *button_text;
  ClutterColor white = { 0xff, 0xff, 0xff, 0xff };
  ClutterColor black = { 0x00, 0x00, 0x00, 0xff };
  gfloat width, height;

  button = clutter_actor_new ();

  button_bg = clutter_actor_new ();
  clutter_actor_set_background_color (button_bg, &white);
  clutter_actor_add_child (button, button_bg);
  clutter_actor_set_opacity (button_bg, 0xcc);

  button_text = clutter_text_new_full ("Sans 10", text, &black);
  clutter_actor_add_child (button, button_text);
  clutter_actor_get_size (button_text, &width, &height);

  clutter_actor_set_size (button_bg, width + PADDING * 2, height + PADDING * 2);
  clutter_actor_set_position (button_bg, 0, 0);
  clutter_actor_set_position (button_text, PADDING, PADDING);

  return button;
}

#define TILE_SQUARE_SIZE 64

static gboolean
draw_background_tile (ClutterCanvas *canvas,
    cairo_t *cr,
    int width,
    int height)
{
  cairo_pattern_t *pat;
  gint no_of_squares_x = width / TILE_SQUARE_SIZE;
  gint no_of_squares_y = height / TILE_SQUARE_SIZE;
  gint row, column;

  /* Create the background tile */
  pat = cairo_pattern_create_linear (width / 2.0, 0.0, width, height / 2.0);
  cairo_pattern_add_color_stop_rgb (pat, 0, 0.662, 0.662, 0.662);
  cairo_set_source (cr, pat);
  cairo_rectangle (cr, 0, 0, width, height);
  cairo_fill (cr);
  cairo_pattern_destroy (pat);

  /* Filling the tile */
  cairo_set_source_rgb (cr, 0.811, 0.811, 0.811);
  cairo_set_line_cap (cr, CAIRO_LINE_CAP_ROUND);

  for (row = 0; row < no_of_squares_y; ++row)
    {
      for (column = 0; column < no_of_squares_x; column++)
        {
          /* drawing square alternatively */
          if ((row % 2 == 0 && column % 2 == 0) ||
              (row % 2 != 0 && column % 2 != 0))
            cairo_rectangle (cr,
                column * TILE_SQUARE_SIZE,
                row * TILE_SQUARE_SIZE,
                TILE_SQUARE_SIZE,
                TILE_SQUARE_SIZE);
        }
      cairo_fill (cr);
    }
  cairo_stroke (cr);
  
  return TRUE;
}


int
main (int argc,
    char *argv[])
{
  ClutterActor *actor, *stage, *buttons, *button;
  ChamplainMarkerLayer *layer;
  ChamplainPathLayer *path;
  gfloat width, total_width = 0;

  if (clutter_init (&argc, &argv) != CLUTTER_INIT_SUCCESS)
    return 1;

  stage = clutter_stage_new ();
  clutter_actor_set_size (stage, 800, 600);
  g_signal_connect (stage, "destroy", G_CALLBACK (clutter_main_quit), NULL);

  /* Create the map view */
  actor = champlain_view_new ();
  clutter_actor_set_size (CLUTTER_ACTOR (actor), 800, 600);
  clutter_actor_add_child (stage, actor);

  /* Create the buttons */
  buttons = clutter_actor_new ();
  clutter_actor_set_position (buttons, PADDING, PADDING);

  button = make_button ("Zoom in");
  clutter_actor_add_child (buttons, button);
  clutter_actor_set_reactive (button, TRUE);
  clutter_actor_get_size (button, &width, NULL);
  total_width += width + PADDING;
  g_signal_connect (button, "button-release-event",
      G_CALLBACK (zoom_in),
      actor);

  button = make_button ("Zoom out");
  clutter_actor_add_child (buttons, button);
  clutter_actor_set_reactive (button, TRUE);
  clutter_actor_set_position (button, total_width, 0);
  clutter_actor_get_size (button, &width, NULL);
  g_signal_connect (button, "button-release-event",
      G_CALLBACK (zoom_out),
      actor);

  clutter_actor_add_child (stage, buttons);

  ClutterContent *canvas;
  canvas = clutter_canvas_new ();
  clutter_canvas_set_size (CLUTTER_CANVAS (canvas), 512, 256);
  g_signal_connect (canvas, "draw", G_CALLBACK (draw_background_tile), NULL);
  clutter_content_invalidate (canvas);
  champlain_view_set_background_pattern (CHAMPLAIN_VIEW (actor), canvas);

  /* Create the markers and marker layer */
  layer = create_marker_layer (CHAMPLAIN_VIEW (actor), &path);
  champlain_view_add_layer (CHAMPLAIN_VIEW (actor), CHAMPLAIN_LAYER (layer));

  /* Connect to the click event */
  clutter_actor_set_reactive (actor, TRUE);
  g_signal_connect (actor, "button-release-event",
      G_CALLBACK (map_view_button_release_cb),
      actor);

  /* Finish initialising the map view */
  g_object_set (G_OBJECT (actor), "zoom-level", 12,
      "kinetic-mode", TRUE, NULL);
  champlain_view_center_on (CHAMPLAIN_VIEW (actor), 45.466, -73.75);

  clutter_actor_show (stage);
  clutter_main ();

  return 0;
}

File: demos/launcher.js
#!/usr/bin/env gjs
/*
 * Copyright (C) 2010 Simon Wenner <simon@wenner.ch>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/**
 * Champlain launcher example in Javascript.
 *
 * Dependencies:
 *  * gobject-introspection
 *  * Build Champlain with '--enable-introspection'
 *  * Gjs: https://gitlab.gnome.org/GNOME/gjs/wikis/Home
 *
 * If you installed Champlain in /usr/local you have to run:
 * export GI_TYPELIB_PATH=$GI_TYPELIB_PATH:/usr/local/lib/girepository-1.0/
 */

const Lang = imports.lang;
const Clutter = imports.gi.Clutter;
const Champlain = imports.gi.Champlain;

Clutter.init (null);

const PADDING = 10;

function make_button (text)
{
  let white = new Clutter.Color ({red:0xff, blue:0xff, green:0xff, alpha:0xff});
  let black = new Clutter.Color ({red:0x00, blue:0x00, green:0x00, alpha:0xff});

  let button = new Clutter.Actor ();

  let button_bg = new Clutter.Actor ();
  button_bg.set_background_color (white)
  button.add_child (button_bg);
  button.set_opacity (0xcc);

  let button_text = new Clutter.Text ({font_name:"Sans 10", text:text, color:black});
  button.add_child (button_text);

  let [width, height] = button_text.get_size();
  button_bg.set_size (width + PADDING * 2, height + PADDING * 2);
  button_bg.set_position (0, 0);
  button_text.set_position (PADDING, PADDING);

  return button;
}

function map_view_button_release_cb (actor, event)
{
  // FIXME: it does not print and GLib.printf does not work.
  if (event.button != 1 || event.click_count > 1)
    return false;

  let lat = view.y_to_latitude (event.y);
  let lon = view.x_to_longitude (event.x);

  log ("Map clicked at %f, %f \n", lat, lon);

  return true;
}

let stage = new Clutter.Stage ();
stage.title = "Champlain Javascript Example";
stage.set_size (800, 600);

/* Create the map view */
let view = new Champlain.View();
view.set_size (800, 600);
stage.add_actor (view);

/* Create the buttons */
let buttons = new Clutter.Actor ();
let total_width = 0;
buttons.set_position (PADDING, PADDING);

let button = make_button ("Zoom in");
buttons.add_child (button);
button.set_reactive (true);

let width = button.width;
total_width += width + PADDING;
button.connect ("button-release-event", Lang.bind (view,
    function (actor, event)
      {
        view.zoom_in ();
        return true;
      }));

let button = make_button ("Zoom out");
buttons.add_child (button);
button.set_reactive (true);
button.set_position (total_width, 0);

let width = button.width;
total_width += width + PADDING;
button.connect ("button-release-event", Lang.bind (view,
    function (actor, event)
      {
        view.zoom_out ();
        return true;
      }));

stage.add_child (buttons);

/* Create the markers and marker layer */
let orange= Clutter.Color.new(0xf3,0x94,0x07,0xbb);
let layer=new Champlain.MarkerLayer();
let marker=Champlain.Label.new_with_text("Sample Marker","Serif 14",null,orange);
marker.set_location(45.466, -73.75);
layer.add_marker(marker);
marker.set_reactive(true);
layer.show();
view.add_layer(layer);

/* Connect to the click event */
view.set_reactive (true);
view.connect ("button-release-event", Lang.bind (view,
    map_view_button_release_cb));

/* Finish initialising the map view */
view.zoom_level = 12;
view.kinetic_mode = true;
view.center_on (45.466, -73.75);

stage.connect ("destroy", Clutter.main_quit);

stage.show ();
Clutter.main ();
